File: backend/app/noc.py
"""NOC REST APIs; no extension to Protocol v1.0 messages."""
from fastapi import APIRouter, HTTPException
from starlette.concurrency import run_in_threadpool
from .pt_controller import controller_from_environment
from .network_agent import NetworkAgent

router = APIRouter(prefix="/api", tags=["NOC"])
agent = NetworkAgent(controller=controller_from_environment())


@router.get("/network/state")
async def network_state():
    return await run_in_threadpool(agent.snapshot)


@router.get("/network/events")
async def network_events():
    return {"source": "SIMULATED", "events": agent.event_history()}


from typing import Literal
from pydantic import BaseModel, ConfigDict

class SecurityRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    event: Literal["PORT_SECURITY_VIOLATION", "ACL_BLOCK_EVENT"] = "PORT_SECURITY_VIOLATION"

@router.get("/security/state")
async def security_state():
    return agent.security_snapshot()

@router.post("/simulation/security")
async def security_attack(request: SecurityRequest):
    return agent.security_attack(request.event)

@router.post("/simulation/security/restore")
async def security_restore():
    return agent.restore_security()


cloud_failed = False
sync_status = "NOT_RUN"

class BranchRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    device: Literal["R-BRANCH", "SW-BRANCH"]
    source_ip: str = "192.168.30.20"

@router.post("/branch/check")
async def branch_check(request: BranchRequest):
    return agent.check_device(request.device, request.source_ip)

@router.get("/branch/state")
async def branch_state():
    baseline = agent.provider.get_state()["management"]
    return {"source": "SIMULATED", "allowed_source": baseline["allowed_source"],
            "devices": [agent.check_device(device, baseline["allowed_source"]) for device in baseline["devices"]]}

@router.get("/campus-policy")
async def campus_policy():
    from . import main
    baseline = agent.provider.get_state()["campus_policy"]
    return {**baseline, "campus_version": f"campus-{baseline['version']} / thermal-v{main.state.policy['version']}",
            "thermal": dict(main.state.policy), "source": "Edge policy live; network/security configuration overlay",
            "edge_ack": "See existing Policy ACK; desired policy alone does not prove application"}

@router.post("/simulation/network")
async def network_failure():
    if agent.controller is not None:
        raise HTTPException(409, "真实控制器模式不允许用模拟故障覆盖观测结果")
    return agent.set_network_failure(True)

@router.post("/simulation/network/restore")
async def network_restore():
    if agent.controller is not None:
        raise HTTPException(409, "真实控制器模式没有模拟网络故障需要恢复")
    return agent.set_network_failure(False)

@router.get("/simulation/state")
async def simulation_state():
    from . import main
    return {"cloud": "OFFLINE" if cloud_failed else "ONLINE", "cloud_failed": cloud_failed,
            "edge_connected": main.state.edge_online,
            "edge_mode": "AUTONOMOUS MODE (expected)" if cloud_failed and main.state.policy["mode"] == "AUTO" else main.state.policy["mode"],
            "fan": main.state.fan_state, "fan_observation": "LAST KNOWN · live telemetry unavailable" if cloud_failed or not main.state.edge_online else "LIVE EDGE TELEMETRY",
            "state_sync": sync_status, "network_failed": agent.network_failed,
            "source": "Control-channel outage drill; HTTP backend stays available"}

@router.post("/simulation/cloud")
async def cloud_failure():
    global cloud_failed, sync_status
    from . import main
    if not cloud_failed:
        cloud_failed = True
        sync_status = "WAITING_FOR_RECOVERY"
        socket = main.edge_socket
        main.edge_socket = None
        main.state.mark_edge_offline()
        await main.broadcast_snapshot()
        if socket is not None:
            await socket.close(code=1013, reason="Cloud failure drill")
        agent.record("CLOUD", "CONTROL_CHANNEL_DOWN", "Real Edge WebSocket closed; HTTP recovery remains available")
    return await simulation_state()

@router.post("/simulation/cloud/restore")
async def cloud_restore():
    global cloud_failed, sync_status
    if cloud_failed:
        cloud_failed = False
        sync_status = "WAITING_FOR_STATE_SYNC"
        agent.record("CLOUD", "CONTROL_CHANNEL_RESTORED", "Awaiting actual Edge reconnect and state_sync")
    return await simulation_state()

@router.get("/noc/state")
async def noc_state():
    return {"network": await run_in_threadpool(agent.snapshot), "security": agent.security_snapshot(),
            "branch": await branch_state(), "policy": await campus_policy(),
            "simulation": await simulation_state(), "events": agent.event_history()}


@router.get("/controller/state")
async def controller_state():
    if agent.controller is None:
        return {"configured": False, "status": "NOT_CONFIGURED", "devices": [], "topology": None,
                "source": "PT_CONTROLLER", "error": "设置 PT_CONTROLLER_URL、PT_CONTROLLER_USERNAME、PT_CONTROLLER_PASSWORD 后重启 Backend"}
    return await run_in_threadpool(agent.controller.snapshot)

File: dashboard/noc.js
/* NOC adapter state is independent of the existing Edge WebSocket. */
(() => {
  const el = (id) => document.getElementById(id);
  let busy = false;
  function status(id, value, healthy) {
    const node = el(id);
    node.textContent = value;
    node.className = `noc-status ${healthy ? "good" : ["UNAVAILABLE", "UNKNOWN", "NOT_CONFIGURED", "NOT COLLECTED"].includes(value) ? "unknown" : "bad"}`;
  }
  function renderNetwork(state) {
    ["nocOspf", "nocBgp", "nocTunnel"].forEach(id => status(id, "NOT COLLECTED", false));
    el("networkSourceBadge").textContent = "PT 控制器 · 实际数据";
    el("networkFailure").disabled = true;
    el("networkRestore").disabled = true;
    renderController(state.source && state.source.kind === "PT_CONTROLLER" ? state.controller : null);
  }
  function renderController(controller) {
    const target = el("controllerDevices");
    target.textContent = "";
    el("networkDeviceCards").textContent = "";
    if (!controller || !controller.configured || controller.source !== "PT_CONTROLLER") {
      status("controllerStatus", "NOT_CONFIGURED", false);
      el("controllerNotice").textContent = "尚未接入控制器。配置 PT_CONTROLLER_URL、用户名和密码后重启 Backend；网络健康仅接收真实 NC 数据，不显示模拟状态。";
      el("nocNetworkNotice").textContent = "尚未配置真实 NC；没有设备健康数据。";
      el("controllerTime").textContent = "";
      el("controllerTopology").textContent = "暂无控制器数据";
      return;
    }
    status("controllerStatus", controller.status, controller.status === "CONNECTED");
    const errors = {MISSING_CREDENTIALS:"未设置控制器用户名或密码",AUTH_FAILED:"控制器认证失败，请检查账户",CONNECTION_FAILED:"无法连接控制器，请检查 Real World Access 和端口",INVALID_INVENTORY_RESPONSE:"设备清单响应格式不兼容",INVALID_AUTH_RESPONSE:"登录响应中没有有效票据",INVALID_JSON:"控制器返回了无效 JSON"};
    el("controllerNotice").textContent = controller.error ? `${errors[controller.error] || controller.error}；未使用模拟数据替代。` : `实际读取 ${controller.devices.length} 台设备 · ${controller.url}${controller.devices.length ? "" : " · 清单为空，请先执行设备发现"}${controller.topology_error ? " · 拓扑暂不可用" : ""}`;
    el("controllerTime").textContent = `采集时间：${controller.observed_at || "未成功采集"} · 后端最多每 5 秒查询一次`;
    el("nocNetworkNotice").textContent = el("controllerNotice").textContent;
    const devices = controller.status === "CONNECTED" && controller.source === "PT_CONTROLLER" ? controller.devices : [];
    devices.forEach(device => {
      const name = device.hostname || device.name || device.id || "未命名";
      const ip = device.managementIpAddress || device.ipAddress || "未提供";
      const collection = device.collectionStatus || "NOT COLLECTED";
      const online = collection === "Managed";
      const card = document.createElement("article"); card.className = "network-device-card";
      [["h3", name], ["p", `管理 IP：${ip}`], ["p", `控制器状态：${collection}`], ["b", online ? "ONLINE" : collection]].forEach(([tag, value]) => {
        const node = document.createElement(tag); node.textContent = String(value);
        if (tag === "b") node.className = `noc-status ${online ? "good" : "unknown"}`;
        card.appendChild(node);
      });
      el("networkDeviceCards").appendChild(card);
      const row = document.createElement("tr");
      [device.hostname || device.name || device.id || "未命名", device.managementIpAddress || device.ipAddress || "未提供", device.type || device.deviceType || device.family || "未提供", online ? "ONLINE · Managed" : collection].forEach(value => {
        const cell = document.createElement("td"); cell.textContent = String(value); row.appendChild(cell);
      });
      target.appendChild(row);
    });
    el("controllerTopology").textContent = controller.topology ? JSON.stringify(controller.topology, null, 2) : "暂无可用拓扑数据";
  }
  function securityDetail(event) {
    return event.event === "PORT_SECURITY_VIOLATION" ? "检测到未经授权的 MAC · 端口已阻断" : event.event === "ACL_BLOCK_EVENT" ? "检测到未授权流量 · ACL 已阻断" : event.detail;
  }
  const display = value => ({"AUTONOMOUS MODE (expected)":"本地自治模式（预期）", "NOT_RUN":"尚未演练", "WAITING_FOR_RECOVERY":"等待恢复云端", "WAITING_FOR_STATE_SYNC":"等待 Edge 状态同步", "SUCCESS":"同步成功 SUCCESS", "LAST KNOWN · live telemetry unavailable":"最后观测值 · 当前无法获取实时遥测", "LIVE EDGE TELEMETRY":"Edge 实时遥测"}[value] || value);
  function renderSecurity(state) {
    status("nocSecurity", state.port_security, state.port_security === "SECURE");
    el("nocViolations").textContent = state.violations;
    el("nocAcl").textContent = state.acl;
    el("nocPort").textContent = state.port_status;
    const card = el("nocSecurityEvent");
    card.textContent = state.last_event ? `${state.last_event.event} · ${securityDetail(state.last_event)}${state.active ? "" : " · 已恢复"}` : "暂无安全事件";
    card.className = state.active ? "security-alert" : "notice";
  }
  async function securityAction(path) {
    try {
      const response = await fetch(path, {method: "POST", headers: {"Content-Type": "application/json"}, body: "{}", signal: AbortSignal.timeout(5000)});
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const data = await response.json();
      renderSecurity(data.security || data);
      el("securityActionResult").textContent = "演练状态已更新";
    } catch (error) { el("securityActionResult").textContent = `演练失败： ${error.message}`; }
  }
  el("securityAttack").addEventListener("click", () => securityAction("/api/simulation/security"));
  el("securityRestore").addEventListener("click", () => securityAction("/api/simulation/security/restore"));
  function renderCenters(data) {
    const policy = data.policy;
    el("campusVersion").textContent = policy.campus_version;
    el("campusThermal").textContent = `${policy.thermal.policy_id} · v${policy.thermal.version} · ${policy.thermal.threshold_c}℃ · ${policy.thermal.mode}`;
    el("campusNetwork").textContent = `分部访问 ${policy.network.branch_access} · IoT 隔离 ${policy.network.iot_isolation ? "已启用" : "未启用"}`;
    el("campusSecurity").textContent = `端口安全 ${policy.security.port_security}`;
    data.branch.devices.forEach(device => { el(device.device === "R-BRANCH" ? "routerStatus" : "switchStatus").textContent = device.status; });
    const simulation = data.simulation;
    status("simCloud", simulation.cloud, simulation.cloud === "ONLINE");
    el("simMode").textContent = simulation.edge_connected ? display(simulation.edge_mode) : simulation.cloud_failed ? display(simulation.edge_mode) : "边缘节点未连接";
    el("simFan").textContent = simulation.fan;
    el("simSync").textContent = display(simulation.state_sync);
    el("fanObservation").textContent = display(simulation.fan_observation);
  }
  async function action(button, path, payload, output) {
    button.disabled = true;
    el(output).textContent = "正在执行…";
    try {
      const response = await fetch(path, {method: "POST", headers: {"Content-Type": "application/json"}, body: JSON.stringify(payload), signal: AbortSignal.timeout(5000)});
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const result = await response.json();
      el(output).textContent = result.management ? `${result.device} · 远程管理 ${result.management} · VTY ACL ${result.acl}` : "演练已执行 · 正在刷新状态";
      await refresh();
    } catch (error) { el(output).textContent = `操作失败： ${error.message}`; }
    finally { button.disabled = false; }
  }
  const bindings = [
    ["checkRouter", "/api/branch/check", {device: "R-BRANCH"}, "branchResult"],
    ["checkSwitch", "/api/branch/check", {device: "SW-BRANCH"}, "branchResult"],
    ["cloudFailure", "/api/simulation/cloud", {}, "simulationResult"],
    ["cloudRestore", "/api/simulation/cloud/restore", {}, "simulationResult"],
    ["networkFailure", "/api/simulation/network", {}, "simulationResult"],
    ["networkRestore", "/api/simulation/network/restore", {}, "simulationResult"],
    ["simAttack", "/api/simulation/security", {}, "simulationResult"],
    ["simSecurityRestore", "/api/simulation/security/restore", {}, "simulationResult"],
  ];
  bindings.forEach(([id, path, payload, output]) => el(id).addEventListener("click", () => action(el(id), path, payload, output)));
  async function refresh() {
    if (busy) return;
    busy = true;
    try {
      const response = await fetch("/api/noc/state", {cache: "no-store", signal: AbortSignal.timeout(10000)});
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      const data = await response.json();
      renderNetwork(data.network);
      renderSecurity(data.security);
      renderCenters(data);
    } catch (error) {
      ["nocOspf", "nocBgp", "nocTunnel"].forEach(id => status(id, "NOT COLLECTED", false));
      el("networkDeviceCards").textContent = "";
      ["nocSecurity", "nocViolations", "nocAcl", "nocPort", "routerStatus", "switchStatus", "campusVersion", "campusThermal", "campusNetwork", "campusSecurity", "simCloud", "simMode", "simFan", "simSync"].forEach(id => { el(id).textContent = "UNAVAILABLE"; });
      status("controllerStatus", "UNAVAILABLE", false);
      el("controllerNotice").textContent = "Backend 不可用，无法确认控制器最新状态";
      el("controllerDevices").textContent = "";
      el("controllerTopology").textContent = "最新数据不可用";
      status("nocSecurity", "UNAVAILABLE", false);
      status("simCloud", "UNAVAILABLE", false);
      el("nocSecurityEvent").textContent = "安全状态不可用；等待新数据";
      el("nocSecurityEvent").className = "notice";
      el("fanObservation").textContent = "运维接口不可用；无法确认实时状态";
      el("nocNetworkNotice").textContent = "Network Agent 不可用；正在重试，未沿用旧健康状态。";
    } finally { busy = false; }
  }
  refresh();
  setInterval(refresh, 1500);
})();
